lock-api/src/main/java/com/palantir/lock/watch/ConjureLockRequestMetadataUtils.java

```diff
@@ -0,0 +1,168 @@
+import com.palantir.lock.watch.ChangeMetadata.Created;
+import com.palantir.lock.watch.ChangeMetadata.Deleted;
+import com.palantir.lock.watch.ChangeMetadata.Unchanged;
+import com.palantir.lock.watch.ChangeMetadata.Updated;
+import com.palantir.logsafe.Unsafe;
+import com.palantir.util.IndexEncodingUtils;
+import com.palantir.util.IndexEncodingUtils.ChecksumType;
+import com.palantir.util.IndexEncodingUtils.IndexEncodingResult;
+import com.palantir.util.IndexEncodingUtils.KeyListChecksum;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.immutables.value.Value;
+
+public final class ConjureLockRequestMetadataUtils {
+    private ConjureLockRequestMetadataUtils() {}
+
+    @VisibleForTesting
+    static final ChecksumType DEFAULT_CHECKSUM_TYPE = ChecksumType.CRC32_OF_DETERMINISTIC_HASHCODE;
+
+    public static ConjureMetadataConversionResult toConjureIndexEncoded(
+            Set<LockDescriptor> lockDescriptors, LockRequestMetadata metadata) {
+        IndexEncodingResult<LockDescriptor, ConjureChangeMetadata> encoded = IndexEncodingUtils.encode(
+                lockDescriptors,
+                metadata.lockDescriptorToChangeMetadata(),
+                changeMetadata -> changeMetadata.accept(ChangeMetadataToConjureVisitor.INSTANCE),
+                DEFAULT_CHECKSUM_TYPE);
+        KeyListChecksum checksum = encoded.keyListChecksum();
+        ConjureLockRequestMetadata conjureLockRequestMetadata = ConjureLockRequestMetadata.builder()
+                .indexToChangeMetadata(encoded.indexToValue())
+                .lockListChecksum(checksumToConjure(checksum))
+                .build();
+        return ImmutableConjureMetadataConversionResult.builder()
+                .lockList(encoded.keyList())
+                .conjureMetadata(conjureLockRequestMetadata)
+                .build();
+    }
+
+    public static LockRequestMetadata fromConjureIndexEncoded(ConjureMetadataConversionResult conversionResult) {
+        List<LockDescriptor> keyList = conversionResult.lockList();
+        ConjureLockRequestMetadata conjureMetadata = conversionResult.conjureMetadata();
+        KeyListChecksum checksum =
+                checksumFromConjure(conversionResult.conjureMetadata().getLockListChecksum());
+        IndexEncodingResult<LockDescriptor, ConjureChangeMetadata> encoded =
+                IndexEncodingResult.<LockDescriptor, ConjureChangeMetadata>builder()
+                        .keyList(keyList)
+                        .indexToValue(conjureMetadata.getIndexToChangeMetadata())
+                        .keyListChecksum(checksum)
+                        .build();
+        Map<LockDescriptor, Optional<ChangeMetadata>> optChangeMetadata = IndexEncodingUtils.decode(
+                encoded,
+                conjureChangeMetadata -> conjureChangeMetadata.accept(ChangeMetadataFromConjureVisitor.INSTANCE));
+        // visitUnknown() will return an empty optional
```

Comment:
non-blocking/super-nit: I don't think this comment is needed. If people are curious as to why `Optional` is being returned, I'd expect them to go read `ChangeMetadataFromConjureVisitor`